x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractor.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.datafeed.extractor.aggregation;
+
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils;
+import org.elasticsearch.xpack.core.rollup.action.RollupSearchAction;
+
+/**
+ * An implementation that extracts data from elasticsearch using search with aggregations against rollup indexes on a client.
+ * The first time {@link #next()} is called, the search is executed. The result aggregations are
+ * stored and they are then processed in batches. Cancellation is supported between batches.
+ * Note that this class is NOT thread-safe.
+ */
+class RollupDataExtractor extends AbstractAggregationDataExtractor<RollupSearchAction.RequestBuilder> {
+
+    RollupDataExtractor(Client client, AggregationDataExtractorContext dataExtractorContext) {
+        super(client, dataExtractorContext);
+    }
+
+    @Override
+    protected RollupSearchAction.RequestBuilder buildSearchRequest() {
+        // For derivative aggregations the first bucket will always be null
+        // so query one extra histogram bucket back and hope there is data
+        // in that bucket
+        long histogramSearchStartTime = Math.max(0, context.start - ExtractorUtils.getHistogramIntervalMillis(context.aggs));
```

Comment:
Most of this method is still common between the 2 implementations. There are some flaky bits here, like the -1 histogram bucket for the derivative aggregations. Would be nice to keep those in a single place.

What if we make the abstract method by `buildSearchRequest(SearchSourceBuilder)`, then we have the common logic in the super class? We can set the types in the `AggregationDataExtractor` while we don't do so in the rollup?